### PR TITLE
#3054. Add covariant parameters tests

### PR DESCRIPTION
**Pull request by sgrekhov**
If this PR is Ok I'll add the same tests for mixins, extension types etc in the next one.

### CODE REVIEW

**Review by eernstg — COMMENTED** [review]
Looks good!

I commented on a couple of tests because I think they could be stronger if they would use an actual type argument for `X` which is different from a top type: This means that we can see a failure if there is a bug where the type of a formal parameter is relaxed to `Object?`, but it was already `Object?` because that was the value of `X`.

On the other hand, this is of course a corner case, too, so maybe it's fine to have a test where `X` is bound to `Object?`.

However, it's more important to ensure that the relaxation of the formal parameter type to `Object?` does take place than it is to check that nothing wrong happens when that relaxation is a no-op (unless this is already covered by some other test).

File: Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t02.dart [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import '../../../../Utils/static_type_helper.dart';
+
+class C<X> {
+  X m<Y extends X>(X r1, Y r2, {required X p1}) {
+    return 42 as X;
+  }
+}
+
+main() {
+  C o = C();
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t02.dart]: Updated

File: Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t02.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  final f = o.m;
+  f.expectStaticType<
+        Exactly<Object? Function<Y extends Object?>(Object? r1, Y r2, {required Object? p1})>
+      >();
+
+  Expect.isTrue(
+    f is Object? Function<Y extends Object?>( // ignore: unnecessary_type_check
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t02.dart:47]: But analyzer does produce this warning. And there is no "binding of `X` to `String`" in this test. May be the next test is meant (and there is no this warning there)?

File: Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t03.dart [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+///
+/// @description Check that if `pj` is covariant then `Tj` is `Object?` at
+/// run time.
+/// @author sgrekhov22@gmail.com
+
+import '../../../../Utils/expect.dart';
+import '../../../../Utils/static_type_helper.dart';
+
+class C<X extends num> {
+  X m<Y extends String>(
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t03.dart]: Updated

File: Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t03.dart [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    covariant X r1,
+    covariant Y r2, {
+    required covariant int p1,
+  }) {
+    return 42 as X;
+  }
+}
+
+main() {
+  C o = C();
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t03.dart]: Updated

File: Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_positional_params_A03_t02.dart [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  C o = C();
+  final f = o.m;
+  f.expectStaticType<
+        Exactly<
+          Object? Function<Y extends Object?>(Object? r1, Y r2, [Object? p1])
+        >
+      >();
+
+  Expect.isTrue(
+    f is Object? Function<Y extends Object?>(Object? r1, Y r2, [Object? p1]) // ignore: unnecessary_type_check
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_positional_params_A03_t02.dart:50]: Updated here as well

**Review by sgrekhov — COMMENTED** [review]
Thanx! Updated. PTAL

File: Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t02.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (eernstg) [reply Language/Expressions/Property_Extraction/Instance_Method_Closurization/method_closurization_named_params_A03_t02.dart]: Thx!

**Review by eernstg — APPROVED** [review]
Thanks, LGTM!
